maven-compat/src/main/java/org/apache/maven/repository/legacy/DefaultWagonManager.java

```diff
@@ -465,6 +469,28 @@ public void getRemoteFile(
         }
     }
 
+    private void copyFile(File source, File destination) throws IOException {
+        String message;
+        if (!source.exists()) {
+            message = "File " + source + " does not exist";
+            throw new IOException(message);
+        } else if (!source.getCanonicalPath().equals(destination.getCanonicalPath())) {
+            File parentFile = destination.getParentFile();
+            if (parentFile != null && !parentFile.exists()) {
+                parentFile.mkdirs();
+            }
+            Files.copy(
+                    source.toPath(),
+                    destination.toPath(),
+                    StandardCopyOption.REPLACE_EXISTING,
+                    StandardCopyOption.COPY_ATTRIBUTES);
+            if (source.length() != destination.length()) {
```

Comment:
This if block isn't needed. If Files.copy doesn't copy everything, an IOException will be thrown above.

Reply:
I have no idea what could happen in case of concurrent access to the repository...
Copy is not necessarily atomic, so I'm not sure that an exception will always be thrown before this check.

Reply:
Even with this check it's not safe against concurrent modifications outside this thread or process. I'm willing to assume the JDK method behaves as documented until proven otherwise. And inlining questionable code is no better than writing it from scratch. This block should go.

Reply:
maven-compat is deprecated and should be gone in Maven 4.1, there's no point in enhancing the code there.  Quite the opposite, any change may introduce regressions in legacy / untested code.  Untested, because all the ITs have now been updated and that code path isn't used anymore, but in very plugins.  If you want to modify a code that has been proven to work, please provide a unit test for that then.

Also, the following test simply exhibits that the check has an effect and is not a no-op.  The documentation does in no way guarantee that the copy is atomic, it's actually specifically written that it's not.  I agree there's no way simple way to ensure that in this code, and I certainly would not want to go in that direction for the reason that code is bound to be removed soon.
```
    public static void main(String[] args) throws Exception {
        Path dir = Files.createTempDirectory("test-");

        Path source = dir.resolve("source.txt");
        try (Writer w = Files.newBufferedWriter(source, StandardCharsets.UTF_8)) {
            for (int i = 0; i < 1024 * 1024; i++) {
                w.append("01234567890\n");
            }
        }

        Path dest = dir.resolve("dest.txt");
        for (int t = 0; t < 4; t++) {
            new Thread(() -> {
                try {
                    Files.copy(source, dest, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.COPY_ATTRIBUTES);
                    if (Files.size(source) != Files.size(dest)) {
                        throw new IOException("Failed to copy full contents from " + source + " to " + dest);
                    }
                } catch (IOException e) {
                    throw new IOError(e);
                }
            }).start();
        }
    }
```

Reply:
Fwiw, maven-shared-utils' version of it also has the same check: https://github.com/apache/maven-shared-utils/blob/master/src/main/java/org/apache/maven/shared/utils/io/FileUtils.java#L760

Reply:
@gnodet
I find the copyFile method of maven-shared-utils deprecated, I think we don't need to use it anymore.

Reply:
There's some shared history between all of these, and they have the same failure modes and issues. That's why we need to move away from all of them and use the JDK methods that are available in Java 7+ that did not exist 20 years ago when this buggy code was first written.